src/stashcache.py

```diff
@@ -234,6 +241,109 @@ def generate_public_cache_authfile(vo_data, resource_groups, fqdn=None, legacy=T
     return authfile
 
 
+def generate_cache_scitokens(vo_data: VOsData, resource_groups: List[ResourceGroup], fqdn: str, suppress_errors=True) -> str:
+    """
+    Generate the SciTokens needed by a StashCache cache server, given the fqdn
+    of the cache server.
+
+    The scitokens config for a StashCache namespace is in the VO YAML and looks like:
+
+        DataFederations:
+            StashCache:
+                Namespaces:
+                    /store:
+                        - SciTokens:
+                            Issuer: https://scitokens.org/cms
+                            Base Path: /
+                            Restricted Path: /store
+
+    "Restricted Path" is optional.
+    `fqdn` must belong to a registered cache resource.
+
+    If suppress_errors is True, returns an empty string on various error conditions (e.g. no fqdn,
+    no resource matching fqdn, resource does not contain a cache server, etc.).  Otherwise, raises
+    ValueError or DataError.
+
+    """
+    template = """\
+[Global]
+audience = {resource.name}, https://{fqdn}
+
+{issuer_blocks_str}
+"""
+
+    if not fqdn:
+        if suppress_errors:
+            return ""
+        else:
+            raise ValueError("fqdn: empty")
+
+    resource = _get_cache_resource(fqdn, resource_groups, suppress_errors)
+    if not resource:
+        return ""
+
+    issuer_blocks = []
+    for vo_name, vo_data in vo_data.vos.items():
+        stashcache_data = vo_data.get('DataFederations', {}).get('StashCache')
+        if not stashcache_data:
+            continue
+
+        namespaces = stashcache_data.get("Namespaces", {})
+        if not namespaces:
+            continue
+
+        has_non_public = False
```

Comment:
Oof.  The `not has_non_public` double-negative later on is hard on the brain.

How about calling this `needs_authz`?  Then it's obvious what 

```
if not needs_authz:
    continue
```

is doing later on.

Reply:
Sure!